### DISCUSSION TITLE
ST7567 Which constructor to use among three that work ok?

### DISCUSSION THREAD

**Discussion by olonsoft** [olikraus/u8g2#2272] [Q&A] [answered]
I am using this ST7567 board.
![IMG20230927081948](https://github.com/olikraus/u8g2/assets/17516317/06027f50-3ec0-44ab-b939-900d9360abff)
![IMG20230927081934](https://github.com/olikraus/u8g2/assets/17516317/8a696965-e55e-413a-aeb7-2a8dfbb8ed97)

I can successfully use it with all of the following constructors on esp8266 with `setI2CAddress(0x7e)`
```txt
U8G2_ST7567_ENH_DG128064I_F_HW_I2C   // with contrast about 230
U8G2_ST7567_OS12864_F_HW_I2C         // with contrast about 25
U8G2_ST7567_JLX12864_F_HW_I2C        // with contrast about 150
```
What is the difference among these constructors and what is the best to choose for my project?
Thank you.

**Comment by olikraus** [comment]
If the content is displayed correctly, then the difference is only the init code which was given for the display as mentioned in the constructor.
I mean the display vendor for the JLX12864 has provided a different init sequence compared to the OS12864 display. We probably would need to look into the code itself to see the difference.

Indeed the difference between the JLX and the OS is just the v0 setup and the contrast setting.
JLX12864:
https://github.com/olikraus/u8g2/blob/f50cb994708eb1e63c2efcc68685a08f7fcc988f/csrc/u8x8_d_st7567.c#L286-L287

OS12864:
https://github.com/olikraus/u8g2/blob/f50cb994708eb1e63c2efcc68685a08f7fcc988f/csrc/u8x8_d_st7567.c#L1074-L1075

I would say: If a constructor works, then just use that constructor.

**Reply by olikraus** [reply, marked as answer]
The controller might be the same, but still every display has its own setup sequence. This init sequence has to be provided by the display vendor, in this case maybe https://goldenmorninglcd.com/graphic-lcd-display/128x64-st7567-gm12864-03b-white-on-blue/ 

However it might be difficult to get this information and in many cases a slighty deviating init sequence will also work, so... you can just use what fits best.

**Comment by muyunqingge** [comment]
Hello, i also use  st7567 to drive the screen. My device environment is arm-linux. I try to use the func `u8g2_Setup_st7567_i2c_jlx12864_f`, but my st7567 CMD_BYTE is 0x80, DATA_BYTE is 0xC0. On my terminal , i find the u8g2 is work, but the CMD_DATA and DATA_BYTE it sends is 0x0 and 0x40. Pls, how i should modify it. 
This is my code, i use the example about arm-linux.
![图片](https://github.com/olikraus/u8g2/assets/82813546/f9cdb195-c902-4372-8481-717fe6cba4f8)
This is my terminal print.
![图片](https://github.com/olikraus/u8g2/assets/82813546/81d211b7-f2b1-4f76-919e-4ff742da5a87)

Thank you very much!
